core/src/test/scala/com/stripe/dagon/RealNumberTest.scala

```diff
@@ -827,12 +838,63 @@ class RealNumberTest extends FunSuite {
   }
 
   test("optimization does not change evaluation") {
-    val genMap = Gen.mapOf(Gen.zip(Gen.choose('a', 'z').map(_.toString), Gen.choose(-1000.0, 1000.0)))
-
-    forAll(Real.genReal(3), genMap)  { (r, vars) =>
-      val optR = optimizeAll(r)
+    def law(r: Real, vars: Map[String, Double], ruleSeq: Seq[Rule[RealN]]) = {
+      val optR = Dag.applyRuleSeq[Any, RealN](r, toLiteral, ruleSeq)
       closeOpt(optR.evaluate(vars), r.evaluate(vars), s"$optR, $r")
     }
+
+     val ruleSeqGen: Gen[Seq[Rule[RealN]]] =
+       implicitly[Arbitrary[Set[Rule[RealN]]]].arbitrary.map(_.toSeq)
+
+     val genMap = Gen.mapOf(Gen.zip(Gen.choose('a', 'z').map(_.toString), Gen.choose(-1000.0, 1000.0)))
+
+     def genCompleteMap(r: Real): Gen[Map[String, Double]] = {
+       val vars = r.freeVars
+       val sz = vars.size
+       Gen.listOfN(sz, Gen.choose(-1000.0, 1000.0)).map { ds =>
+         vars.zip(ds).toMap
+       }
+     }
+
+     val completeEval =
+       Real
+         .genReal(3)
+         .flatMap { r => genCompleteMap(r).map((r, _)) }
+
+    forAll(Real.genReal(3), genMap, ruleSeqGen)(law(_, _, _))
+
+    forAll(completeEval,
```

Comment:
this scalacheck found the errors much faster since it should always return a value. This could very easily find the bugs.